### ISSUE TITLE
Mixins as composable building blocks?

### ISSUE THREAD

**Issue by chikamichi** [inuitcss/inuitcss#305]
Hi,

I would like to get feedback from core maintainers about the following idea :)

Given that `@extend` is to be avoided for good reasons, inuitcss, much like most CSS framework, ends up "having" developers couple their HTML and CSS through the percolation of imperative micro-classes. Those classes function at different level of abstraction/specificity so they may not feel like imperative *per se*, but they are fundamentally imperative in that, at their own level of action, they dictate a small chunk of behaviour or "step" into styling the element as a whole, and it is only through their combination that a tag truly gains a full-fledged "UI intent".

CONTRIBUTING.md highlights a goal of simplicity for inuitcss:

> inuitcss is written in Sass, which provides us with a lot of incredibly powerful features. However, we do not want inuitcss to become a platform for Sass’ capabilities—all code that comes into inuitcss should be as simple and CSS-like as possible.

But it also says that inuitcss is about

> team-based workflows, long-running and larger projects, and […] flexibility.

I have come to feel like parsing through these spreading combinations of imperative micro-classes in the HTML, and being forced to perform a constant mental map/reduce of styles, runs counter to those goals.

What if inuitcss as a framework would do better to promote simplicity in its usage over its implementation, by consenting to adding just a tad of "complexity" in the form of an abstraction layer: mixins.

What I'd like to be able to do, is limit my HTML/CSS coupling to only my declarative (by design!) BEM declarations, or whatever naming/namespacing convention is in use. That is, *in my stylesheets* rather than in `class=` declarations, I would like to be able to mix-in / compose imperative builders like .o-block, .u-whatever, etc. into a declarative, business-relevant .c-mycomponent namespace (just an example).

Note that it does *not* fundamentally alter the inner workings of the framework: classes from different & consistent level of abstractions/specificity still get combined to form a broader rendering, but coupling between the structure and its styling has been greatly reduced, down to BEM's initial intent: naming things in the HTML, implementing in the stylesheets.

In order to allow for that composition of building blocks in the CSS rather than the HTML, inuitcss would need to encapsulate its core styles into reusable mixins (eg. `@mixin o-block`, so that one may `@include o-block`). A rather extremist stance would be that of *not* exposing those mixins through micro-classes so as to be as little tantalizing as possible, but of course it limits usage to Sass code-bases and goes against simplicity of use, so the smart thing to do is to actually expose classes (`.o-block { @include o-block; }`).

Given the small footprint of inuitcss, it seems like it would not be that big of a deal or added complexity for a huge win in terms of flexibility, reusability, uncoupling and ownership. Of course it's entirely possible doing so in our own code bases, but that's tedious and that's why I'm asking about promoting the concept to the framework.

Has that concept been around, discussed and/or rejected already? Why? I was unable to find insights in the issues or PR :)

Thank you!

**Comment by csshugs** [comment]
@chikamichi I truly can see your point and understand the idea behind your approach. It is—objectively—neither better nor worse than how we currently working with inuitcss. 

But what it is is one thing: A huge difference in the way inuitcss was and is meant to be used. It's not about how difficult it would be to introduce this concept into the framework—it would be work, but it definitely would be manageable. 

It's rather exactly that concept of putting together pieces of your UI by sticking together different classes right in the markup that we are after and how we want inuitcss to be used. 

As I said, it would definitely be possible to introduce your proposed concept into inuitcss, but then _I_ would probably stop using inuitcss, and I'm sure the rest of @inuitcss/core is thinking similar about this. To emphasize this: This isn't about making the usual compromises you got to make if you are maintaining a somewhat popular open source project like inuitcss. I did and still do not like every change we introduced due to popular demand of the community. But that's Ok! These were relatively small compromises based popular demands from the community. This however is just something else. 

I'm aware that the way to work with inuitcss would not change, if we would still provide all the classes like this:

```scss
.o-block {
  @include o-block();
}
```

But this would increase the complexity “behind the scenes”.

---

Thanks for your detailed proposal. As I said, it's neither the better nor the worse way of approaching frontend architecture. It's just fundamentally different from our current approach and therefore I'm against the implementation of this idea.
